### DISCUSSION TITLE
OutOfMemoryError: GC overhead limit exceeded compling for Android (Animate)

### DISCUSSION THREAD

**Discussion by smartcodegames** [airsdk/Adobe-Runtime-Support#696] [Q&A] [answered]
I believe the java compiler heap size must be increased. How I do that in Animate?

**Comment by ajwfrost** [comment]
I believe Animate is invoking the ADT command directly rather than using a command-line.. normally I would say adjust the command line with an "-Xmx" parameter but for this, it may be that you can set an environment variable `JAVA_TOOL_OPTIONS` with this parameter?

See
https://stackoverflow.com/questions/1565388/increase-heap-size-in-java
https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/envvars002.html

The settings for -Xmx should be passed through to any tool invocations so e.g. running ADT with values for heap size should mean that when it launches 'javac' it will provide the same arguments to this..

Hope that helps!
thanks

**Reply by smartcodegames** [reply]
I tried modifying adt.bat but probably you were right and adt is executed directly so there was no change. I also tried:
set JAVA_TOOL_OPTIONS="-Xmx4096m"
but only the a error message changed from "OutOfMemoryError: GC overhead limit exceeded" to "OutOfMemoryError: Java heap space". I'm not sure if it's a substantial difference...

**Comment by marchbold** [comment]
Hi,

I don't use animate much anymore, but there used to be a `jvm.ini` file that you could edit to change the java options.

I just had a quick look in my install and there is one in the application directory and one in the settings, both have `-Xmx256m` in them, so potentially try modifying that file?

```
~/Library/Application Support/Adobe/Animate 2021/en_US/Configuration/ActionScript 3.0/jvm.ini
```

**Reply by smartcodegames** [reply]
I found two files with that name:
\Program Files\Adobe Animate CC 2019\Common\First Run\ActionScript 3.0\jvm.ini
\Users\Smartcode\AppData\Local\Adobe\Animate CC 2019\en_US\Configuration\ActionScript 3.0\jvm.ini

Modified both, still the same error :(   I'm getting desperate...

**Reply by marchbold** [reply]
Try adding `-Xms2048m` as well. That sets the minimum heap size.

**Reply by smartcodegames** [reply]
Same result, but adt.log showed some clues (see below)

**Reply by itlancer** [reply]
@smartcodegames, have you tried to restart Animate after jvm.ini edit? Animate apply such new changes only after restart.

**Reply by smartcodegames** [reply]
Yes, multiple times.

**Comment by ajwfrost** [comment, marked as answer]
Hi - if you're able to get it building outside of Animate, it might be easier to solve. If you edit the "adt.cfg" file that's in your SDK's "lib" folder, can you add a line: `DebugOut=true` and then run it again from Animate? Then in your home folder you should end up with a file called "adt.log" which will tell you the parameters that are being used to invoke ADT. This should help you run it from the command-line...

Then you can manually add -Xmx onto it, and other such commands..

**Reply by smartcodegames** [reply]
Thanks, at last some clues: in adt.log there is a line starting with:
DX tool command line: java -Xmx1024m -jar D:\Program Files\Adobe Animate CC 2019\AIR33.1\lib\android\bin\dx.jar --dex --multi-dex ........
So dx.jar should be configured with more heap. But where?

**Reply by ajwfrost** [reply]
This tool is adding an 'Xmx' value based upon the arguments that are provided to the 'parent' i.e. the initial JVM when it's launched. So if using Animate it would be their own Java classes that kick off the ADT activity; if you run ADT from the command-line then you can pass in whatever values you want here and they'd get passed through. Did you also ad 'Xms' to the Animate configuration because that's another one that should get passed through? If either of those are found, we don't add our 'default' value which would be 1024 which is looks like is happening here...

BTW is it definitely this part where the failure comes from? What do you actually see in Animate, or in the ADT log, after the failure? If DX fails, there should be "Unable to run dx" or "dx tool failed", or maybe "failed to create: " with a dex filename.

**Reply by ajwfrost** [reply]
@smartcodegames what version of the AIR SDK are you using btw?

**Reply by smartcodegames** [reply]
It's AIR33.1 (latest), tried with earlier versions too, same error. The error message is here:

DX tool command line: java -Xmx1024m ..............
ADT exception: dx tool failed:
UNEXPECTED TOP-LEVEL ERROR:
java.lang.OutOfMemoryError: GC overhead limit exceeded

It is followed by some details which changes time to time:

	at java.util.zip.InflaterInputStream.<init>(Unknown Source)
	at java.util.zip.ZipFile$ZipFileInflaterInputStream.<init>(Unknown Source)
	at java.util.zip.ZipFile.getInputStream(Unknown Source)
	at com.android.dx.cf.direct.ClassPathOpener.processArchive(ClassPathOpener.java:271)
	at com.android.dx.cf.direct.ClassPathOpener.processOne(ClassPathOpener.java:165)
	at com.android.dx.cf.direct.ClassPathOpener.process(ClassPathOpener.java:143)
	at com.android.dx.command.dexer.Main.processOne(Main.java:678)
	at com.android.dx.command.dexer.Main.processAllFiles(Main.java:569)
	at com.android.dx.command.dexer.Main.runMultiDex(Main.java:365)
	at com.android.dx.command.dexer.Main.runDx(Main.java:286)
	at com.android.dx.command.dexer.Main.main(Main.java:244)
	at com.android.dx.command.Main.main(Main.java:95)

**Reply by ajwfrost** [reply]
Okay thanks - so yes, definitely a DX failure..

If you've got the SDK 33.1.1.406, can you try downloading the below adt.jar file and copying it into your SDK's "lib" folder. Also in that folder you'll find the "adt.cfg" file to which you can add a line "JavaXmx=2048m"
https://transfer.harman.com/message/V2GklnvEgUbMA4uH7EkxaX

Then please let us know what the adt.log output is...

thanks

**Reply by smartcodegames** [reply]
Good news: DX tool command line: java -Xmx2048m ...
So adt.cfg / "JavaXmx=2048m" passed through properly with the new adt.jar

However the build did not succeed because my jre is too old (copied from CC2014). I use it because my certificate (old 1024bit) is not accepted by the new sdk-s (I get UsageError: unable to load a certificate in this file). But it goes to another thread, thanks for your effort!

**Reply by ajwfrost** [reply]
Okay good - that configuration setting is something I just added to see if would help :-) but we can push that into the next release.

If you've got 1024-bit certificates, you need to use Java 1.7, or the first Java 1.8 release which still supports them.
